src/systems/userinput/userinput.js

```diff
@@ -22,36 +21,123 @@ import { viveUserBindings } from "./bindings/vive-user";
+    if (satisfiesPath(binding, path)) {
+      return true;
+    }
+  }
+  return false;
+};
+
+const satisfiedBy = (binding, bindings) => {
+  for (const path of Object.values(binding.src)) {
+    if (path.startsWith("/device/")) continue;
+    if (!satisfyPath(bindings, path)) return false;
+  }
+  return true;
+};
+
+function dependencySort(mappings) {
+  const unsorted = [];
+  for (const mapping of mappings) {
     for (const setName in mapping) {
-      if (!activeSets.has(setName) || !mapping[setName]) continue;
       for (const binding of mapping[setName]) {
-        const { root, priority } = binding;
-        const prevBinding = priorityMap.get(root);
-        if (!root || !priority) {
-          activeBindings.add(binding);
-        } else if (!prevBinding) {
-          activeBindings.add(binding);
-          priorityMap.set(root, binding);
-        } else if (priority > prevBinding.priority) {
-          activeBindings.delete(priorityMap.get(root));
-          activeBindings.add(binding);
-          priorityMap.set(root, binding);
-        } else if (prevBinding.priority === priority) {
-          console.error("equal priorities on same root", binding, priorityMap.get(root));
-        }
+        unsorted.push(binding);
+      }
+    }
+  }
+
+  const sorted = [];
+  while (unsorted.length > 0) {
+    const binding = unsorted.shift();
+    if (satisfiedBy(binding, sorted)) {
+      sorted.push(binding);
+    } else {
+      unsorted.push(binding);
+    }
+  }
+
+  return sorted;
+}
+
+function canMask(masker, masked) {
+  if (masker.priority === undefined) {
+    masker.priority = 0;
+  }
+  if (masked.priority === undefined) {
+    masked.priority = 0;
+  }
+  if (masked.priority >= masker.priority) return false;
+  for (const maskerPath of Object.values(masker.src)) {
+    for (const maskedPath of Object.values(masked.src)) {
+      if (maskedPath.indexOf(maskerPath) !== -1) {
+        return true;
       }
     }
   }
-  return activeBindings;
+  return false;
+}
+
+function computeMasks(bindings) {
+  const masks = [];
+  for (const row in bindings) {
+    for (const col in bindings) {
+      let ColCanMaskRow = false;
+      if (canMask(bindings[col], bindings[row])) {
+        ColCanMaskRow = true;
+      }
+      masks[Number(row) * bindings.length + Number(col)] = ColCanMaskRow;
+    }
+  }
+  return masks;
+}
+
+function isActive(binding, sets) {
+  for (const s of binding.sets) {
+    if (sets.has(s)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+function computeExecutionStrategy(sortedBindings, masks, activeSets) {
+  const actives = [];
+  for (const row in sortedBindings) {
+    actives[row] = isActive(sortedBindings[row], activeSets);
+  }
+
+  const masked = [];
+  for (const row in sortedBindings) {
+    for (const col in sortedBindings) {
+      const rowMask = masked[row] || [];
+      if (masks[Number(row) * sortedBindings.length + Number(col)] && isActive(sortedBindings[col], activeSets)) {
```

Comment:
I'm pretty sure it's considered impolite to iterate over an array with a `for...in` loop. I would write a normal `for` loop.